### PR TITLE
Implement PEP 639, Metadata 2.4

### LINKED ISSUE #16620: Implement PEP 639
**What's the problem this feature will solve?**
Standardized licensing metadata will in time unify the currently largely chaotic landscape.

**Describe the solution you'd like**
See: https://peps.python.org/pep-0639
PEP 639 has been provisionally accepted with one of the conditions being implementation in PyPI.

### PR DESCRIPTION
**Pull request by ewdurbin**
Closes #16620

~~Depends on merge/release of https://github.com/pypa/packaging/pull/828~~
~~TODO: Move to release of packaging with license-expression parser~~

### CODE REVIEW

File: warehouse/forklift/legacy.py
Comment (miketheman) [inline warehouse/forklift/legacy.py:1141]: I'm wondering if [`getmember()`](https://docs.python.org/3/library/tarfile.html#tarfile.TarFile.getmember) might work here? Something like:

```python
with tarfile.open(temporary_filename, "r:gz") as tar:
    for license_file_path in license_files:
        try:
            tar.getmember(license_file_path)
        except KeyError:
             ...
```

This operation also has to scan the entire archive, so I don't know if this is particularly beneficial or not, similar to how you've implemented it in zipfile.

**Review by miketheman — COMMENTED** [review]
Looking cool, and coming along nicely! I commented a bit inside.

One thing I'm curious about - at what point should we be looking to consolidate any file operations where we pop open the file and read stuff from it, so we only have to open/iterate once?

File: warehouse/packaging/models.py
Comment (ewdurbin) [reply warehouse/packaging/models.py]: 01e71e471

File: warehouse/forklift/legacy.py
Comment (ewdurbin) [reply warehouse/forklift/legacy.py]: ecda1a978502c028ec24a90fcb051b85c7222822

**Comment by ewdurbin** [comment]
@befeleme would you be able to take a look at this to review the implementation details?

File: warehouse/templates/includes/packaging/project-data.html
Comment (dnicolodi) [inline warehouse/templates/includes/packaging/project-data.html:205]: I would use just "License" for the user visible string here.

**Comment by befeleme** [comment]
> @befeleme would you be able to take a look at this to review the implementation details?

Yes, hopefully tomorrow.

**Comment by befeleme** [comment]
After reading the code I've got just one question: I can't locate the validation of the `License-Expression` per https://peps.python.org/pep-0639/#add-license-expression-field ("[Python Package Index (PyPI) MUST validate that they contain a valid, case-normalized license expression with valid identifiers and MUST reject uploads that do not"). Or did I miss it?

**Comment by ewdurbin** [comment]
> After reading the code I've got just one question: I can't locate the validation of the `License-Expression` per https://peps.python.org/pep-0639/#add-license-expression-field ("[Python Package Index (PyPI) MUST validate that they contain a valid, case-normalized license expression with valid identifiers and MUST reject uploads that do not"). Or did I miss it?

That's implemented upstream in pypa/packaging, see https://github.com/pypa/packaging/pull/828

**Comment by befeleme** [comment]
Ah, looking at the whole `warehouse/forklift/metadata.py` I see `packaging` validations are applied automatically and no additional changes regarding that are needed here. Makes sense.

File: warehouse/migrations/versions/6ee23f5a6c1b_add_pep_639_columns.py
Comment (ewdurbin) [inline warehouse/migrations/versions/6ee23f5a6c1b_add_pep_639_columns.py:30]: I think this needs a closer look to try to avoid contention when rolled out.

File: warehouse/templates/includes/packaging/project-data.html
Comment (ewdurbin) [reply warehouse/templates/includes/packaging/project-data.html:205]: I'm waiting for a more thorough review from @pypi/warehouse-committers before I address suggestions/questions.

**Review by miketheman — APPROVED** [review]
Yay, we're almost there! I didn't re-review a lot of what I'd already covered, so I focused on the migration and testing, and looks like everything generally conforms to the PEP.

I'm thinking that tracking usage/progress of this can be deferred, but that might be a nice metric to emit via DB query later - I can imagine wanting to see new projects using the syntax, older projects converting to the new syntax, etc - but definitely beyond scope of this changeset.

Approving to you can decide when you want to run the migration, but please include some timeouts based on some guesswork from TestPyPI

File: warehouse/forklift/legacy.py
Comment (di) [inline warehouse/forklift/legacy.py:1122]: Seems like the check for `metadata_version` is redundant here (and elsewhere), I don't think `meta.license_files` will exist if the version is < 2.4:

```
>>> Metadata.from_raw({"metadata_version": "2.3", "license_files": ["something"], "version": "6.6.6", "name": "throwaway"})
  + Exception Group Traceback (most recent call last):
  |   File "<stdin>", line 1, in <module>
  |   File "/home/di/.pyenv/versions/3.12.7/lib/python3.12/site-packages/packaging/metadata.py", line 752, in from_raw
  |     raise ExceptionGroup("invalid metadata", exceptions)
  | ExceptionGroup: invalid metadata (1 sub-exception)
  +-+---------------- 1 ----------------
    | packaging.metadata.InvalidMetadata: license-file introduced in metadata version 2.4, not 2.3
    +------------------------------------

>>> Metadata.from_raw({"metadata_version": "2.3", "version": "6.6.6", "name": "throwaway"}).license_files is None
True
```

File: warehouse/templates/includes/packaging/project-data.html
Comment (di) [reply warehouse/templates/includes/packaging/project-data.html:205]: Agreed, we should distinguish between the old "License" and the new "License-Expression" here.

File: warehouse/forklift/metadata.py
Comment (di) [inline warehouse/forklift/metadata.py:248]: Seems like this is a check that should be happening in pypa/packaging instead, but isn't?

```
>>> Metadata.from_raw({"metadata_version": "2.4", "license": "MIT", "license_expression": "MIT", "version": "6.6.6", "name": "throwaway"})
<packaging.metadata.Metadata object at 0x7f44b9647680>
```

Comment (ewdurbin) [reply warehouse/forklift/metadata.py:248]: That validation isn't backed into packaging as the [PEP](https://peps.python.org/pep-0639/#deprecate-license-field) says "tools MAY issue a warning" and "PyPI MUST reject"

File: warehouse/forklift/legacy.py
Comment (ewdurbin) [reply warehouse/forklift/legacy.py:1067]: It does. a8dd46958

File: warehouse/migrations/versions/6ee23f5a6c1b_add_pep_639_columns.py
Comment (ewdurbin) [reply warehouse/migrations/versions/6ee23f5a6c1b_add_pep_639_columns.py:45]: FWIW I think that the exclusive lock point is ultimately our "problem", since we perform reads with `READ COMMITTED` in a transaction.

File: warehouse/migrations/versions/6ee23f5a6c1b_add_pep_639_columns.py
Comment (ewdurbin) [reply warehouse/migrations/versions/6ee23f5a6c1b_add_pep_639_columns.py:52]: 74b7b8cde

File: warehouse/forklift/metadata.py
Comment (ewdurbin) [reply warehouse/forklift/metadata.py:248]: I'm not sure if I agree or disagree here, mainly because I am not certain where the responsibility belongs.

This specific check does ensure that PyPI's implementation meets the specification that

> For all newly-uploaded [distribution archive](https://packaging.python.org/en/latest/glossary/#term-Distribution-Archive)s that include a License-Expression field, the [Python Package Index (PyPI)](https://pypi.org/) MUST reject any that specify both License and License-Expression fields.

No matter the underlying validation, so I'm prone to remain explicit and keep it.

File: warehouse/forklift/legacy.py
Comment (ewdurbin) [reply warehouse/forklift/legacy.py:1124]: 8c2859ae686a674eab04b543899984395be4abed

**Comment by ewdurbin** [comment]
Re @miketheman:

> I'm thinking that tracking usage/progress of this can be deferred, but that might be a nice metric to emit via DB query later - I can imagine wanting to see new projects using the syntax, older projects converting to the new syntax, etc - but definitely beyond scope of this changeset.

That is all technically captured in the immutable release metadata, so aside from watching a pot boil in real-time in datadog I agree adoption metrics are well beyond scope.

File: warehouse/forklift/metadata.py
Comment (di) [reply warehouse/forklift/metadata.py:248]: I think we can leave this for now and figure out later if it should belong upstream!

**Review by di — APPROVED** [review]

**Comment by ewdurbin** [comment]
Migrations are going out for this _ahead_ of merge during this window to avoid unnecessary disruption: http://status.python.org/incidents/q5q5347n3s2b

**Comment by ofek** [comment]
Thanks everyone!

**Pull request merged**
